File: worker/src/parser.ts
// Port dari py-service/parser.py — parsing Excel histori pekerjaan pesanan.
import * as XLSX from "xlsx";

const MAIN_SHEET = "Histori Pekerjaan Pesanan";

const NUMERIC_COLS = [
  "pengeluaran_alokasi",
  "pengeluaran_biaya",
  "pengeluaran_qty",
  "penyelesaian_alokasi",
  "penyelesaian_biaya",
  "penyelesaian_qty",
  "total_alokasi",
  "total_biaya",
  "total_qty",
];

const WANT: Record<string, string> = {
  tanggal: "tanggal",
  "batch no": "batch_no",
  kode: "kode",
  "bahan dan biaya": "bahan_biaya",
  keterangan: "keterangan",
  "pengeluaran barang biaya": "pengeluaran_biaya",
  "pengeluaran barang kuantitas": "pengeluaran_qty",
  "pengeluaran barang alokasi": "pengeluaran_alokasi",
  "penyelesaian pesanan biaya": "penyelesaian_biaya",
  "penyelesaian pesanan kuantitas": "penyelesaian_qty",
  "penyelesaian pesanan alokasi": "penyelesaian_alokasi",
  "total tipe transaksi biaya": "total_biaya",
  "total tipe transaksi kuantitas": "total_qty",
  "total tipe transaksi alokasi": "total_alokasi",
};

const REQUIRED_KEYS = [
  "tanggal",
  "batch no",
  "kode",
  "bahan dan biaya",
  "pengeluaran barang biaya",
  "penyelesaian pesanan biaya",
  "penyelesaian pesanan kuantitas",
];

export type PrdRow = Record<string, any>;

function _norm(v: unknown): string {
  if (v === null || v === undefined) return "";
  return String(v).replace(/\s+/g, " ").trim();
}

function _key(name: string): string {
  return String(name)
    .toLowerCase()
    .replace(/ - /g, " ")
    .replace(/-/g, " ")
    .replace(/ \(%\)/g, "")
    .replace(/\(/g, "")
    .replace(/\)/g, "")
    .replace(/\./g, "");
}

function _num(v: unknown): number {
  if (v === null || v === undefined) return 0;
  const n = Number(v);
  return Number.isNaN(n) ? 0 : n;
}

function _round2(x: number): number {
  return Math.round(x * 100) / 100;
}

type Grid = Array<Array<unknown>>;

function buildGrid(ws: XLSX.WorkSheet): Grid {
  const range = ws["!ref"] ? XLSX.utils.decode_range(ws["!ref"]) : { s: { r: 0, c: 0 }, e: { r: 0, c: 0 } };
  const nR = range.e.r + 1;
  const nC = range.e.c + 1;
  const grid: Grid = Array.from({ length: nR }, () => new Array(nC).fill(undefined));
  const merges = (ws["!merges"] as Array<{ s: { r: number; c: number }; e: { r: number; c: number } }>) ?? [];
  for (let r = 0; r < nR; r++) {
    for (let c = 0; c < nC; c++) {
      const cell = ws[XLSX.utils.encode_cell({ r, c })];
      grid[r][c] = cell ? cell.v : undefined;
    }
  }
  for (const m of merges) {
    if (m.s.r <= 1) {
      const anchor = grid[m.s.r]?.[m.s.c];
      if (anchor !== undefined && anchor !== null) {
        for (let r = m.s.r; r <= Math.min(m.e.r, 1); r++) {
          for (let c = m.s.c; c <= m.e.c; c++) grid[r][c] = anchor;
        }
      }
    }
  }
  return grid;
}

function _flattenColumns(grid: Grid, nC: number): string[] {
  const cols: string[] = [];
  for (let c = 0; c < nC; c++) {
    const rawA = grid[0]?.[c];
    const rawB = grid[1]?.[c];
    const a = String(rawA === undefined || rawA === null ? "nan" : rawA).trim();
    const b = String(
      rawB === undefined || rawB === null ? `Unnamed: ${c}_level_1` : rawB,
    ).trim();
    if (b.startsWith("Unnamed") || b === "") cols.push(a);
    else if (a && a !== b && !a.startsWith("Unnamed")) cols.push(`${a} - ${b}`);
    else cols.push(b);
  }
  return cols;
}

function _singleHeader(grid: Grid, nC: number): string[] {
  const cols: string[] = [];
  for (let c = 0; c < nC; c++) {
    const v = grid[0]?.[c];
    cols.push(v === undefined || v === null ? `Unnamed: ${c}` : String(v));
  }
  return cols;
}

function _mapColumns(columns: string[]): { mapping: Record<string, string>; missing: string[] } {
  const lookup: Record<string, string> = {};
  for (const c of columns) {
    const k = _key(c);
    if (k && !(k in lookup)) lookup[k] = c;
  }
  const mapping: Record<string, string> = {};
  for (const [key, out] of Object.entries(WANT)) {
    if (key in lookup) mapping[out] = lookup[key];
  }
  const missing = REQUIRED_KEYS.filter((c) => !(c in lookup));
  return { mapping, missing };
}

function fmtDate(v: unknown): string | null {
  if (v instanceof Date && !Number.isNaN(v.getTime())) {
    const y = v.getFullYear();
    const m = String(v.getMonth() + 1).padStart(2, "0");
    const d = String(v.getDate()).padStart(2, "0");
    return `${y}-${m}-${d}`;
  }
  if (typeof v === "number") {
    const dc = XLSX.SSF.parse_date_code(v);
    if (dc) return `${dc.y}-${String(dc.m).padStart(2, "0")}-${String(dc.d).padStart(2, "0")}`;
    return null;
  }
  if (typeof v === "string") {
    const s = v.trim();
    let m = /^(\d{4})-(\d{1,2})-(\d{1,2})/.exec(s);
    if (m) return `${m[1]}-${m[2].padStart(2, "0")}-${m[3].padStart(2, "0")}`;
    m = /^(\d{1,2})\/(\d{1,2})\/(\d{4})/.exec(s);
    if (m) return `${m[3]}-${m[2].padStart(2, "0")}-${m[1].padStart(2, "0")}`;
    m = /^(\d{1,2})-(\d{1,2})-(\d{4})/.exec(s);
    if (m) return `${m[3]}-${m[2].padStart(2, "0")}-${m[1].padStart(2, "0")}`;
    const d = new Date(s);
    if (!Number.isNaN(d.getTime())) return fmtDate(d);
    return null;
  }
  return null;
}

function _isEmpty(v: unknown): boolean {
  return v === undefined || v === null || _norm(v) === "";
}

export function parseExcel(content: ArrayBuffer, filename: string) {
  const wb = XLSX.read(new Uint8Array(content), { type: "array", cellDates: true });
  const sheetName = wb.SheetNames.includes(MAIN_SHEET) ? MAIN_SHEET : wb.SheetNames[0];
  if (!sheetName) throw new Error("File Excel kosong (tidak ada sheet).");
  const ws = wb.Sheets[sheetName];
  const grid = buildGrid(ws);
  const nC = grid[0]?.length ?? 0;

  let mapping: Record<string, string>;
  let missing: string[];
  let dataStart: number;
  const first = _mapColumns(_flattenColumns(grid, nC));
  if (first.missing.length === 0) {
    mapping = first.mapping;
    missing = [];
    dataStart = 2;
  } else {
    const second = _mapColumns(_singleHeader(grid, nC));
    if (second.missing.length === 0) {
      mapping = second.mapping;
      missing = [];
      dataStart = 1;
    } else {
      throw new Error(`Kolom berikut tidak ditemukan: ${first.missing.join(", ")}`);
    }
  }

  const colIdx: Record<string, number> = {};
  for (const [out, colName] of Object.entries(mapping)) {
    colIdx[out] = _flattenColumns(grid, nC).indexOf(colName);
  }

  const recs: PrdRow[] = [];
  for (let r = dataStart; r < grid.length; r++) {
    const rec: PrdRow = {};
    for (const key of Object.keys(mapping)) {
      const idx = colIdx[key];
      rec[key] = idx >= 0 ? grid[r]?.[idx] : undefined;
    }
    for (const c of NUMERIC_COLS) {
      if (rec[c] === undefined || rec[c] === null) rec[c] = 0;
    }
    recs.push(rec);
  }

  const before = recs.length;
  let invalidRows = 0;

  const out = recs.filter((r) => !_norm(r.batch_no).toUpperCase().startsWith("TOTAL"));
  const out2 = out.filter((r) => _norm(r.kode) !== "");
  const out3 = out2.filter((r) => {
    const batchNa = _isEmpty(r.batch_no);
    const bb = _norm(r.bahan_biaya).toUpperCase();
    const isFooter =
      batchNa && bb.includes("BIAYA") && _num(r.pengeluaran_biaya) === 0 && _num(r.penyelesaian_biaya) === 0;
    return !isFooter;
  });
  invalidRows += before - out3.length;

  let lastBatch: unknown;
  let lastTanggal: unknown;
  let lastKet: unknown;
  for (const r of out3) {
    if (_isEmpty(r.batch_no)) r.batch_no = lastBatch;
    else lastBatch = r.batch_no;
    if (_isEmpty(r.tanggal)) r.tanggal = lastTanggal;
    else lastTanggal = r.tanggal;
    if (r.keterangan === undefined || _isEmpty(r.keterangan)) r.keterangan = lastKet;
    else lastKet = r.keterangan;
  }

  const out4 = out3.filter((r) => _norm(r.batch_no) !== "");

  let badDate = 0;
  for (const r of out4) {
    const d = fmtDate(r.tanggal);
    if (d === null) badDate++;
    r.tanggal = d;
  }
  invalidRows += badDate;
  const out5 = out4.filter((r) => r.tanggal !== null);

  for (const r of out5) {
    r.total_biaya = _num(r.pengeluaran_biaya) + _num(r.penyelesaian_biaya);
    r.total_qty = _num(r.pengeluaran_qty) + _num(r.penyelesaian_qty);
    r.total_alokasi = _num(r.pengeluaran_alokasi) + _num(r.penyelesaian_alokasi);
  }

  for (const r of out5) {
    for (const f of ["bahan_biaya", "keterangan", "batch_no", "kode"]) {
      r[f] = _norm(r[f]);
    }
    for (const f of NUMERIC_COLS) {
      r[f] = _round2(_num(r[f]));
    }
  }

  const rows = out5;
  const totalBiaya = rows.reduce((a, r) => a + _num(r.total_biaya), 0);
  const dates = rows.map((r) => String(r.tanggal)).filter((d) => d !== "");

  const summary = {
    filename,
    sheet: sheetName,
    row_count: rows.length,
    batch_count: new Set(rows.map((r) => String(r.batch_no))).size,
    kode_count: new Set(rows.map((r) => String(r.kode))).size,
    date_min: dates.length ? dates.reduce((a, b) => (a < b ? a : b)) : "",
    date_max: dates.length ? dates.reduce((a, b) => (a > b ? a : b)) : "",
    total_biaya: _round2(totalBiaya),
  };

  return {
    rows,
    summary,
    missing_columns: [],
    invalid_rows: invalidRows,
  };
}

File: py-service/parser.py
"""Parsing & normalisasi file Excel histori pekerjaan pesanan PT Hijrah Gizihew.

Mendukung header flat maupun 2 tingkat (merged) dari export Accurate.
Menghasilkan kolom sesuai PRD (production_transactions) tanpa modifikasi data raw.
"""

import io
import re

import numpy as np
import pandas as pd

pd.set_option("future.no_silent_downcasting", True)

MAIN_SHEET = "Histori Pekerjaan Pesanan"

REQUIRED = [
    "Tanggal",
    "Batch No.",
    "Kode",
    "Bahan dan Biaya",
    "Pengeluaran Barang - Biaya",
    "Pengeluaran Barang - Kuantitas",
    "Penyelesaian Pesanan - Biaya",
    "Penyelesaian Pesanan - Kuantitas",
]

# Alokasi dan Total bersifat opsional untuk kompatibilitas; diisi 0 jika tidak ada.
OPTIONAL = [
    "Keterangan",
    "Pengeluaran Barang - Alokasi",
    "Penyelesaian Pesanan - Alokasi",
    "Total Tipe Transaksi - Alokasi",
    "Total Tipe Transaksi - Biaya",
    "Total Tipe Transaksi - Kuantitas",
]

OUTPUT_KEYS = [
    "tanggal", "batch_no", "kode", "bahan_biaya", "keterangan",
    "pengeluaran_alokasi", "pengeluaran_biaya", "pengeluaran_qty",
    "penyelesaian_alokasi", "penyelesaian_biaya", "penyelesaian_qty",
    "total_alokasi", "total_biaya", "total_qty",
]


def _norm(text) -> str:
    if text is None or (isinstance(text, float) and np.isnan(text)):
        return ""
    return re.sub(r"\s+", " ", str(text)).strip()


def _key(name: str) -> str:
    return (
        name.lower()
        .replace(" - ", " ")
        .replace("-", " ")
        .replace(" (%)", "")
        .replace("(", "")
        .replace(")", "")
        .replace(".", "")
    )


def _pick_sheet(xls: pd.ExcelFile) -> str:
    if MAIN_SHEET in xls.sheet_names:
        return MAIN_SHEET
    return xls.sheet_names[0]


def _flatten_columns(reader_columns) -> list:
    """Flatten header[0,1] yang berasal dari cell merged Accurate."""
    cols: list[str] = []
    for pair in reader_columns:
        a, b = str(pair[0]).strip(), str(pair[1]).strip()
        if b.startswith("Unnamed") or b == "":
            cols.append(a)
        elif a and a != b and not a.startswith("Unnamed"):
            cols.append(f"{a} - {b}")
        else:
            cols.append(b)
    return cols


def _map_columns(columns: list) -> tuple[dict, list]:
    """Petakan nama kolom mentah ke key normal; kembalikan (mapping, missing)."""
    mapping: dict[str, str] = {}
    lookup = {}
    for c in columns:
        k = _key(c)
        if k and k not in lookup:
            lookup[k] = c

    want = {
        "tanggal": "tanggal",
        "batch no": "batch_no",
        "kode": "kode",
        "bahan dan biaya": "bahan_biaya",
        "keterangan": "keterangan",
        "pengeluaran barang biaya": "pengeluaran_biaya",
        "pengeluaran barang kuantitas": "pengeluaran_qty",
        "pengeluaran barang alokasi": "pengeluaran_alokasi",
        "penyelesaian pesanan biaya": "penyelesaian_biaya",
        "penyelesaian pesanan kuantitas": "penyelesaian_qty",
        "penyelesaian pesanan alokasi": "penyelesaian_alokasi",
        "total tipe transaksi biaya": "total_biaya",
        "total tipe transaksi kuantitas": "total_qty",
        "total tipe transaksi alokasi": "total_alokasi",
    }
    for key, out in want.items():
        if key in lookup:
            mapping[out] = lookup[key]

    required_keys = ["tanggal", "batch no", "kode", "bahan dan biaya",
                     "pengeluaran barang biaya", "penyelesaian pesanan biaya",
                     "penyelesaian pesanan kuantitas"]
    missing = [c for c in required_keys if c not in lookup]
    return mapping, missing


def _read_sheet(content: bytes):
    xls = pd.ExcelFile(io.BytesIO(content))
    sheet = _pick_sheet(xls)

    # Coba dua tingkat header (export Accurate baru)
    df2 = pd.read_excel(xls, sheet_name=sheet, header=[0, 1])
    cols2 = _flatten_columns(df2.columns)
    df2.columns = cols2
    mapping, missing = _map_columns(cols2)
    if not missing:
        return df2, sheet, mapping

    # Fallback: header satu tingkat (format lama)
    df1 = pd.read_excel(xls, sheet_name=sheet, header=0)
    mapping, missing1 = _map_columns([str(c) for c in df1.columns])
    if not missing1:
        return df1, sheet, mapping
    raise ValueError(f"Kolom berikut tidak ditemukan: {', '.join(missing)}")


def parse_excel(content: bytes, filename: str) -> dict:
    df, sheet, mapping = _read_sheet(content)

    known = {v: k for k, v in mapping.items()}
    df = df.rename(columns=known)
    keep = [k for k in OUTPUT_KEYS if k in df.columns]
    df = df[keep].copy()

    numeric_cols = [
        "pengeluaran_alokasi", "pengeluaran_biaya", "pengeluaran_qty",
        "penyelesaian_alokasi", "penyelesaian_biaya", "penyelesaian_qty",
        "total_alokasi", "total_biaya", "total_qty",
    ]
    for col in numeric_cols:
        if col in df.columns:
            df[col] = pd.to_numeric(df[col], errors="coerce")
        else:
            df[col] = 0.0
    df[numeric_cols] = df[numeric_cols].fillna(0.0)

    # ===== Pembersihan baris subtotal / footer / grand-total =====
    # Kolom Batch No. & Tanggal ter-merge per batch/hari (pandas membaca NaN pada baris lanjutan).
    before = len(df)
    batch_orig = df["batch_no"].astype(str).str.strip().str.upper()

    # 1) Subtotal per batch: "Total Batch No." (teks TOTAL)
    df = df[~batch_orig.str.startswith("TOTAL")]
    # 2) Baris tanpa kode (baris judul/subtotal tak bernama)
    df = df[df["kode"].map(lambda v: _norm(v)) != ""]
    # 3) Footer harian: batch NaN + baris biaya nol (mis. "Biaya Potong" 0)
    footer = (
        df["batch_no"].isna()
        & df["bahan_biaya"].map(lambda v: _norm(v).upper()).str.contains("BIAYA", na=False)
        & (df["pengeluaran_biaya"] == 0)
        & (df["penyelesaian_biaya"] == 0)
    )
    df = df[~footer]

    # 4) Isi batch/tanggal/keterangan yang ter-merge (forward-fill)
    for col in ("batch_no", "tanggal", "keterangan"):
        s = df[col].astype(object)
        empty = s.map(
            lambda v: v is None
            or (isinstance(v, float) and np.isnan(v))
            or str(v).strip() == ""
        )
        df[col] = s.where(~empty, np.nan).ffill()

    invalid_rows = before - len(df)
    df = df[df["batch_no"].notna() & (df["batch_no"].astype(str).str.strip() != "")]

    df["tanggal"] = pd.to_datetime(df["tanggal"], errors="coerce")
    bad_date = df["tanggal"].isna()
    invalid_rows += int(bad_date.sum())
    df = df[~bad_date]

    # Hitung ulang total agar konsisten dengan sisi pengeluaran/penyelesaian
    df["total_biaya"] = df["pengeluaran_biaya"] + df["penyelesaian_biaya"]
    df["total_qty"] = df["pengeluaran_qty"] + df["penyelesaian_qty"]
    df["total_alokasi"] = df["pengeluaran_alokasi"] + df["penyelesaian_alokasi"]

    df["bahan_biaya"] = df["bahan_biaya"].map(_norm)
    df["keterangan"] = df["keterangan"].map(_norm)
    df["batch_no"] = df["batch_no"].map(_norm)
    df["kode"] = df["kode"].map(_norm)
    df["tanggal"] = df["tanggal"].dt.strftime("%Y-%m-%d")

    rows = df[OUTPUT_KEYS].copy()
    records = rows.to_dict(orient="records")
    for r in records:
        for key in numeric_cols:
            r[key] = round(float(r[key]), 2)

    summary = {
        "filename": filename,
        "sheet": sheet,
        "row_count": len(records),
        "batch_count": int(rows["batch_no"].nunique()),
        "kode_count": int(rows["kode"].nunique()),
        "date_min": rows["tanggal"].min() if len(rows) else "",
        "date_max": rows["tanggal"].max() if len(rows) else "",
        "total_biaya": round(float(rows["total_biaya"].sum()), 2),
    }
    return {
        "rows": records,
        "summary": summary,
        "missing_columns": [],
        "invalid_rows": invalid_rows,
    }